### PR TITLE
Listeners not invoked when remote participant times out

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

Suppose a participant dies abruptly.  Eventually, all of the readers
associated with the writers of that participant will become
disassociated.  Readers with listeners should receive
`on_data_available` callbacks where the instance state is not
`NOT_ALIVE_NO_WRITERS_INSTANCE_STATE`.  Currently, this does not
happen.

Solution
--------

When a writer is removed, write a synthetic unregister sample for each
instance in the writer.

### CODE REVIEW

File: dds/DCPS/InstanceState.h [hunk 1]
```diff
@@ -120,6 +120,13 @@ class OpenDDS_Dcps_Export InstanceState : public ReactorInterceptor {
                           int num_alive_writers,
                           const MonotonicTimePoint& when);
 
+  /// Returns true if the writer is a writer of this instances.
```
Comment (mitza-oci) [inline dds/DCPS/InstanceState.h:123]:
```suggestion
  /// Returns true if the writer is a writer of this instance.
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 2]
```diff
@@ -2290,14 +2290,15 @@ DataReaderImpl::writer_became_dead(WriterInfo& info,
...
   ACE_GUARD(ACE_Recursive_Thread_Mutex, instance_guard, this->instances_lock_);
-  for (SubscriptionInstanceMapType::iterator iter = instances_.begin(),
-      next = iter; iter != instances_.end(); iter = next) {
+  for (SubscriptionInstanceMapType::iterator iter = instances_.begin(), next = iter;
+       iter != instances_.end(); iter = next) {
     ++next;
-    iter->second->instance_state_->writer_became_dead(
-        info.writer_id_, liveliness_changed_status_.alive_count, when);
+    if (iter->second->instance_state_->writes_instance(info.writer_id_)) {
+      set_instance_state(iter->first, DDS::NOT_ALIVE_NO_WRITERS_INSTANCE_STATE, SystemTimePoint::now(), info.writer_id_);
```
Comment (mitza-oci) [inline dds/DCPS/DataReaderImpl.cpp:2300]: Compile error here in no-BITs builds

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

File: dds/DCPS/DataReaderImpl.h [hunk 3]
```diff
@@ -529,11 +529,12 @@ class OpenDDS_Dcps_Export DataReaderImpl
     DDS::SampleStateMask sample_states, DDS::ViewStateMask view_states,
     DDS::InstanceStateMask instance_states) = 0;
 
+#endif
+
   virtual void set_instance_state(DDS::InstanceHandle_t instance,
```
Comment (iguessthislldo) [inline dds/DCPS/DataReaderImpl.h:534]: Changes to this method require changes to `tests/DCPS/unit/MyTypeSupportImpl.cpp`.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dds/DCPS/DataReaderImpl.h (hunk 3 above)
Comment (ClaytonCalabrese) [reply dds/DCPS/DataReaderImpl.h:534]: I'm confused what you mean by changes are needed to the cpp? It looks like that function was stubbed out in the .h. Are you saying you want the function tested in Main.cpp?

Comment (iguessthislldo) [reply dds/DCPS/DataReaderImpl.h:534]: Yes, I was saying the stub needed to be changed.

File: tests/DCPS/ManualAssertLiveliness/DataReaderListenerImpl.cpp [hunk 4]
```diff
@@ -39,7 +37,8 @@ void DataReaderListenerImpl::on_data_available(DDS::DataReader_ptr reader)
 
   if (status == DDS::RETCODE_OK) {
     if (si.valid_data == 1) {
-      ACE_DEBUG((LM_INFO,
+      ++num_reads_;
+     ACE_DEBUG((LM_INFO,
```
Comment (iguessthislldo) [inline tests/DCPS/ManualAssertLiveliness/DataReaderListenerImpl.cpp:41]:
```suggestion
      ACE_DEBUG((LM_INFO,
```

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**